applications/FluidDynamicsApplication/custom_elements/data_containers/qs_vms/qs_vms_residual_derivatives.h

```diff
@@ -0,0 +1,562 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+#if !defined(KRATOS_QS_VMS_RESIDUAL_FIRST_DERIVATIVES_H)
+#define KRATOS_QS_VMS_RESIDUAL_FIRST_DERIVATIVES_H
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "containers/variable.h"
+#include "geometries/geometry.h"
+#include "geometries/geometry_data.h"
+#include "includes/constitutive_law.h"
+#include "includes/node.h"
+#include "includes/process_info.h"
+#include "includes/ublas_interface.h"
+#include "utilities/time_discretization.h"
+
+// Application includes
+#include "custom_utilities/fluid_element_utilities.h"
+#include "custom_utilities/fluid_calculation_utilities.h"
+#include "custom_elements/data_containers/qs_vms/qs_vms_derivative_utilities.h"
+
+namespace Kratos
+{
+///@name Kratos Classes
+///@{
+
+template <unsigned int TDim, unsigned int TNumNodes>
+class QSVMSResidualDerivatives
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    using IndexType = std::size_t;
+
+    using NodeType = Node<3>;
+
+    using GeometryType = Geometry<NodeType>;
+
+    using PropertiesType = typename Element::PropertiesType;
+
+    static constexpr IndexType TBlockSize = TDim + 1;
+
+    /// Size of the strain and stress vectors (in Voigt notation) for the formulation
+    constexpr static IndexType TStrainSize = (TDim - 1) * 3; // 3 in 2D, 6 in 3D
+
+    constexpr static IndexType TElementLocalSize = TBlockSize * TNumNodes;
+
+    constexpr static IndexType TNN = TNumNodes;
+
+    using ArrayD = array_1d<double, TDim>;
+
+    using VectorN = BoundedVector<double, TNumNodes>;
+
+    using VectorF = BoundedVector<double, TElementLocalSize>;
+
+    using MatrixDD = BoundedMatrix<double, TDim, TDim>;
+
+    using MatrixND = BoundedMatrix<double, TNumNodes, TDim>;
+
+    ///@}
+    ///@name Static Operations
+    ///@{
+
+    static void Check(
+        const Element& rElement,
+        const ProcessInfo& rProcessInfo);
+
+    static GeometryData::IntegrationMethod GetIntegrationMethod();
+
+    ///@}
+    ///@name Class Declarations
+    ///@{
+
+    class QSVMSResidualData;
```

Comment:
I'd add some comment explaining the purpose of this declaration.

Reply:
done :)